### PR TITLE
Greedy Stageless

### PR DESCRIPTION
**Pull request by hymm**
This is a proposal for an API and algorithm for a stageless scheduler.

[Rendered](https://github.com/hymm/rfcs/blob/greedy-stageless/rfcs/greedy-stageless.md)

### CODE REVIEW

File: rfcs/greedy-stageless.md [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+}
+```
+
+### Exclusive Systems
+
+Exclusive systems are systems that depend on direct world access. They are marked with the `.exclusive()` builder function and only have the `World` parameter. The `label`, `before`, and `after` apis can be used to order exclusive systems and can be mixed with labels for parallel systems.
+
+### Run Criteria
+
+Run criteria determine whether or not a system will run. They now are separated from LoopingCriteria. A RunCriteria system is a system that returns a bool with true for run the dependent system and false for not running the system. Systems without run criteria will be run once per tick.
```
Comment (Nilirad) [inline rfcs/greedy-stageless.md:77]: Compatible with #33 🙂

File: rfcs/greedy-stageless.md [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+        );
+}
+```
+
+### Looping Criteria
+
+A looping criteria is a system that returns a bool. Returning true makes a system reevaluate its run criteria and false means the system is done running for this tick. Adding looping criteria introduces an implicit requirement for a hard sync. The hard sync is required to reevaluate looping criteria and run criteria. Systems scheduled after a looping system will run after the looping criteria returns false.
+
+```rs
+fn loop_twice(loop_count: Local<int32>) {
```
Comment (Nilirad) [inline rfcs/greedy-stageless.md:106]: `i32`, not `int32`. Maybe `usize` is preferable in this case.

You also missed `-> bool`

File: rfcs/greedy-stageless.md [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+
+Stages will be removed as part of this RFC. Not sure on the exact plan yet, but I think it should be straightforward.
+
+Current stages will probably be replaced by system sets. But chould also be redone with more mind payed to the data flow.
+
+## Drawbacks
+
+- This makes a bevy tick a lot less linear and harder to reason about. There will probably need to be good visualization tools to help deal with this.
+- Mapping the current stages onto the new method may get confusing. If a
+  system is in PREUPDATE and another system is in POSTUPDATE what is the relationship between the two? Are they related at all? Are they implicitly depending on something happending in UPDATE? It will take some careful thought to correctly map the current stages and systems to this new api.
```
Comment (Nilirad) [inline rfcs/greedy-stageless.md:195]: A no-brainer solution would be:

```
PREUPDATE -> PREUPDATE(commands) -> UPDATE -> UPDATE(commands) -> POSTUPDATE
```

Maybe not the best solution, but the behavior of the three systems should not change.

File: rfcs/greedy-stageless.md [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+
+#### Some Scenarios
+
+- If _parallel system A_ depends on _parallel system B_ it can be queued to run in a step because parallel system B will be able to run in the current parallel step.
+- If _parallel system A_ depends on _exclusive system B_ that is a direct blocking dependency and cannot run in the current parallel step. It has to wait until system B runs in some exclusive step.
+- It gets a little trickier because if a system has a dependency on a system that has blocking dependency it will also be blocked. i.e. If _parallel system C_ depends on _parallel system A_ which has a dependency on _exclusive system B_, it also cannot be queued to run, because A is blocked by B which blocks C from running. In this scenario a system A will move from a blocking dependency to a non-blocking dependency once C is run.
```
[the remark this answers is not in the document]
Comment (hymm) [reply rfcs/greedy-stageless.md:156]: yeah should probably read 

> non-blocking dependency for C once B is run.

**Comment by alice-i-cecile** [comment]
> I'm not sure how you can get deterministic application of commands without being explicit. i.e. If system A and B can be run in parallel, but system A's commands need to run before system B's.

The `before_if_needed` abstraction may be able to help us out here IMO. We could comfortably overload this to order any commands produced, even if the systems themselves are parallelizable.

**Comment by hymm** [comment]
> E.g. (everything here would become a stage, Frame would just loop)

So this is how I would model your stages with the API I'm thinking about with the caveat that I don't really have a design yet for states or nested system sets.

```rs
enum AppState {
    Idle
    Running
    Done
}

fn main() {
    App::new()
        .add_state(AppState::Idle)
        .add_system(transition_to_running)
        .add_system_set(
            SystemSet::new()
                .label("Startup")
                .with_run_criteria(AppState::Running::on_enter)
        )
        .add_system_set(
            SystemSet::new()
                .label("Frame")
                .with_run_criteria(AppState::Running::on_update)
                .add_system_set(
                    SystemSet::new().commaands_label("First")
                )
                .add_system_set(
                    SystemSet::new().commands_label("PreUpdate").after("First")
                )
                // Update, Render, and Last will be similar to above
                .add_system_set(
                    SystemSet::new()
                        .label("PostUpdate")
                        .after("Update")
                        .add_system_set(
                            SystemSet::new()
                                .label("Audio")
                                .add_system(
                                    audio_system_a.command_label("audio a commands")
                                )
                                .add_system(
                                    audio_system_b.after("audio a commands")
                                )
                        )
                        .add_system_set(
                            SystemSet::new()
                                .label("Transform")
                                .add_system(
                                    parent_update.label("parent update")
                                )
                                .add_system(
                                    transform_propagation.after("parent update")
                                )
                        )
                )
        )
        .add_system_set(
            SystemSet::new()
                .label("Shutdown")
                .with_run_criteria(AppState::Running::on_exit)
        )
}
```

So adding a hard sync point here between audio_system_a and audio_system_b would just be changing from .label to .command_label.

As an aside, I wouldn't organize things like this.  I would just have have `Audio` and `Transform` inside of `Frame` and schedule these system sets against the commands label for things like `Simulation`, `Physics` and `Input`. Also I don't think of this as the final API, but can probably be made a lot more ergonomic with things like the SystemGraph PR.

> so it doesn't matter if the transform systems go in PostUpdate or PostPostUpdate.

> "How can I do more in parallel?" is a design question.

You can also split them up so half of them are in PostUpdate and the other half are in PostPostUpdate. In this example it probably doesn't matter, but in practice this could lead to better parallelism. I really wish I had a better example, but my point was that the user shouldn't think about this at all, whether it's in the design phase or while coding. Changing things about the audio system should have no effect on how you think about transform systems.

> With transitions, the point I wanted to make is that it's not generally safe to make any sort of transition while there are still pending commands.

That makes a lot of sense. It means that updating state, run criteria, and looping criteria should be done between the run of exclusive systems and parallel systems in my 2-step loop.

File: rfcs/greedy-stageless.md [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+- _Removed_ ~~`.stage(Stage)`~~
+- `.exclusive()`
+  - _Removed_ ~~`.at_start()`~~
+  - _Removed_ ~~`.at end()`~~
+
+### Scheduling when to apply commands
+
+Systems with commands have two parts, a parallel part that can be run in parallel with other systems that the user codes and a command queue that needs to be applied to the world during a hard sync point. Since these happen at different times they should be scheduled separately. This is done through three new system descriptor builder functions.
+
+- `.commands_label(Label)` gives the point where the command queue is applied a `Label` which other systems can be scheduled around.
```
[the remark this answers is not in the document]
Comment (Nilirad) [reply rfcs/greedy-stageless.md:47]: After some reasoning, I think manual labeling through `commands_label` is still needed in some cases. The most obvious example is where you give to two or more command queues the same label in order to flush them in the same sync point. I think a form of default labeling could be made though

[the remark this answers is not in the document]
Comment (hymm) [reply rfcs/greedy-stageless.md:47]: That makes sense.  I'll add default labeling to the future work section.

**Comment by alice-i-cecile** [comment]
> We were able to agree that we're mostly just disagreeing on the public API for ordering systems. Where this RFC implicitly schedules sync points, Joy might prefer a more explicit API where you insert systems into hierarchical stages. Expect a competing RFC for that in the future.

I'm largely in agreement with the things said in that conversation :) My suspicion is that we're likely going to need to end up synthesizing the two perspectives, and offering an API to handle both styles.

**Comment by maniwani** [comment]
> Physics, UI, and AI should be able to run in parallel.

It isn't that simple. While they might not have immediate data dependencies, physics and UI systems usually have to maintain a causal relationship. E.g. you want UI to read the most up-to-date component data or you have UI systems that queue events that cause something to happen in physics.

> Long running tasks or expensive computations [such as pathfinding and AI] could be initiated in one frame and blocked in a later one.

Async tasks are pretty much the only solution for long-running computations. Spawn a task, copy/clone the data you need into it (to avoid race conditions), and then have some system poll it for completion later.

Threads should be reserved for things that can always be doing useful work or need rapid polling like rendering, inputs, and audio.

> entities that could be mutually exclusive to that of UI and possibly physics

Data dependencies are defined in terms of component access. You'd need multiple worlds to isolate entities.

**Comment by alice-i-cecile** [comment]
Closing out in favor of #45 for clarity.

**Pull request closed without merging**